File: lc/coreprep/implementation/patterns_reference_implementation/top_k_elements/KClosestNumbersHeap.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KClosestNumbersHeap {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();
        int x = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = kClosestNumbers(a, k, x);
        for (int r : res) {
          System.out.printf("%d ", r);
        }
        System.out.println();
      }
    }
  }

  // O(logn + klogk) / O(k)
  public static List<Integer> kClosestNumbers(int[] a, int k, int x) {
    int n = a.length;

    int closestIdx = binarySearch(a, 0, n - 1, x);
    int low = Math.max(0, closestIdx - k);
    int high = Math.min(closestIdx + k, n - 1);

    PriorityQueue<Integer> minHeap = new PriorityQueue<>(
        (p, q) -> Integer.compare(Math.abs(a[p] - x), Math.abs(a[q] - x)));

    for (int i = low; i <= high; i++) {
      minHeap.add(i);
    }

    List<Integer> res = new ArrayList<>();
    for (int i = 0; i < k; i++) {
      res.add(a[minHeap.poll()]);
    }

    res.sort(Integer::compare);

    return res;
  }

  private static int binarySearch(int[] a, int low, int high, int k) {
    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    if (low > 0) {
      low--;
    }

    return low;
  }
}

File: lc/coreprep/implementation/practice_iterations/top_k_elements_1/k_closest_numbers_heap.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct CompareDiff {
  const vector<int> &a;
  int x;

  CompareDiff(const vector<int> &a, int x) : a(a), x(x) {}

  bool operator()(int p, int q) const { return abs(a[p] - x) < abs(a[q] - x); }
};

int binary_search(const vector<int> &a, int low, int high, int elem) {
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] == elem) {
      return mid;
    } else if (a[mid] < elem) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  if (low > 0) {
    low--;
  }

  return low;
}

// O(logn + klogk) / O(k)
vector<int> k_closest_numbers(const vector<int> &a, int k, int x) {
  int xpos = binary_search(a, 0, a.size() - 1, x);
  int low = max(0, xpos - k);
  int high = min(xpos + k, (int)a.size() - 1);

  CompareDiff comp(a, x);
  priority_queue<int, vector<int>, CompareDiff> min_heap(comp);
  for (int i = xpos; i > low; i--) {
    min_heap.push(i);
  }

  for (int i = xpos + 1; i < high; i++) {
    min_heap.push(i);

    if (min_heap.size() > k) {
      min_heap.pop();
    }
  }

  vector<int> res;
  while (!min_heap.empty()) {
    res.push_back(a[min_heap.top()]);
    min_heap.pop();
  }

  sort(res.begin(), res.end());

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k, x;
  cin >> tt;

  while (tt--) {
    cin >> n >> k >> x;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = k_closest_numbers(a, k, x);
    for (int r : res) {
      cout << r << " ";
    }
    cout << "\n";
  }

  return 0;
}